subprojects/testing-native/src/main/java/org/gradle/nativeplatform/test/xctest/internal/AbstractLocator.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.nativeplatform.test.xctest.internal;
+
+import org.gradle.process.internal.ExecAction;
+import org.gradle.process.internal.ExecActionFactory;
+
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.util.List;
+
+public abstract class AbstractLocator {
+    private final ExecActionFactory execActionFactory;
+    private File cachedLocation;
+
+    protected AbstractLocator(ExecActionFactory execActionFactory) {
+        this.execActionFactory = execActionFactory;
+    }
+
+    protected abstract List<String> getXcrunFlags();
+
+    public File find() {
```

Comment:
that would be ok, a `synchronized` block inside the method would be better (it's more obvious and it also, theoretically, cannot be interfered with by other things synchronizing on this service).